File: solutions/028-find-the-index-of-the-first-occurrence-in-a-string/solution.go
package main

import "fmt"

func strStr(haystack string, needle string) int {
	n, m := len(haystack), len(needle)
	if m == 0 {
		return 0
	}

	for i := 0; i <= n-m; i++ {
		if haystack[i:i+m] == needle {
			return i
		}
	}
	return -1
}

func main() {
	tests := []struct {
		haystack, needle string
		expected         int
	}{
		{"sadbutsad", "sad", 0},
		{"leetcode", "leeto", -1},
		{"hello", "ll", 2},
		{"a", "a", 0},
	}

	for _, tc := range tests {
		result := strStr(tc.haystack, tc.needle)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("%s | haystack=%q, needle=%q => %d (expected %d)\n", status, tc.haystack, tc.needle, result, tc.expected)
	}
}

File: solutions/028-find-the-index-of-the-first-occurrence-in-a-string/solution.py
class Solution:
    def str_str(self, haystack: str, needle: str) -> int:
        n, m = len(haystack), len(needle)
        if m == 0:
            return 0

        for i in range(n - m + 1):
            if haystack[i : i + m] == needle:
                return i
        return -1


if __name__ == "__main__":
    tests = [
        ("sadbutsad", "sad", 0),
        ("leetcode", "leeto", -1),
        ("hello", "ll", 2),
        ("a", "a", 0),
    ]
    sol = Solution()
    for haystack, needle, expected in tests:
        result = sol.str_str(haystack, needle)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | haystack={haystack!r}, needle={needle!r} => {result} (expected {expected})")
